File: packages/agent-mesh/sdks/go/audit_test.go
package agentmesh

import (
	"testing"
)

func TestAuditLogAndVerify(t *testing.T) {
	al := NewAuditLogger()
	e1 := al.Log("agent-1", "data.read", Allow)
	e2 := al.Log("agent-1", "data.write", Deny)

	if e1.PreviousHash != "" {
		t.Error("first entry should have empty PreviousHash")
	}
	if e2.PreviousHash != e1.Hash {
		t.Error("second entry PreviousHash should equal first Hash")
	}
	if !al.Verify() {
		t.Error("chain should be valid")
	}
}

func TestAuditVerifyDetectsTampering(t *testing.T) {
	al := NewAuditLogger()
	al.Log("a", "x", Allow)
	al.Log("a", "y", Deny)

	// tamper
	al.entries[0].AgentID = "tampered"
	if al.Verify() {
		t.Error("chain should be invalid after tampering")
	}
}

func TestAuditGetEntriesFilter(t *testing.T) {
	al := NewAuditLogger()
	al.Log("agent-1", "read", Allow)
	al.Log("agent-2", "write", Deny)
	al.Log("agent-1", "delete", Deny)

	entries := al.GetEntries(AuditFilter{AgentID: "agent-1"})
	if len(entries) != 2 {
		t.Errorf("filtered entries = %d, want 2", len(entries))
	}

	d := Deny
	entries = al.GetEntries(AuditFilter{Decision: &d})
	if len(entries) != 2 {
		t.Errorf("deny-filtered entries = %d, want 2", len(entries))
	}
}

func TestAuditEmptyVerify(t *testing.T) {
	al := NewAuditLogger()
	if !al.Verify() {
		t.Error("empty chain should verify as true")
	}
}

func TestAuditHashesAreUnique(t *testing.T) {
	al := NewAuditLogger()
	e1 := al.Log("a", "action1", Allow)
	e2 := al.Log("a", "action2", Allow)
	if e1.Hash == e2.Hash {
		t.Error("different entries should have different hashes")
	}
}

File: packages/agent-mesh/sdks/typescript/tests/audit.test.ts
import { AuditLogger } from '../src/audit';

describe('AuditLogger', () => {
  let logger: AuditLogger;

  beforeEach(() => {
    logger = new AuditLogger();
  });

  describe('log()', () => {
    it('creates an entry with timestamp and hashes', () => {
      const entry = logger.log({
        agentId: 'agent-1',
        action: 'data.read',
        decision: 'allow',
      });

      expect(entry.timestamp).toBeDefined();
      expect(entry.hash).toHaveLength(64); // SHA-256 hex
      expect(entry.previousHash).toHaveLength(64);
      expect(entry.agentId).toBe('agent-1');
      expect(entry.action).toBe('data.read');
      expect(entry.decision).toBe('allow');
    });

    it('chains entries with previous hashes', () => {
      const first = logger.log({ agentId: 'a', action: 'x', decision: 'allow' });
      const second = logger.log({ agentId: 'b', action: 'y', decision: 'deny' });

      expect(second.previousHash).toBe(first.hash);
    });

    it('first entry uses genesis hash', () => {
      const entry = logger.log({ agentId: 'a', action: 'x', decision: 'allow' });
      expect(entry.previousHash).toBe('0'.repeat(64));
    });
  });

  describe('verify()', () => {
    it('returns true for an empty log', () => {
      expect(logger.verify()).toBe(true);
    });

    it('returns true for a valid chain', () => {
      logger.log({ agentId: 'a', action: 'x', decision: 'allow' });
      logger.log({ agentId: 'b', action: 'y', decision: 'deny' });
      logger.log({ agentId: 'c', action: 'z', decision: 'review' });

      expect(logger.verify()).toBe(true);
    });

    it('detects tampering', () => {
      logger.log({ agentId: 'a', action: 'x', decision: 'allow' });
      logger.log({ agentId: 'b', action: 'y', decision: 'deny' });

      // Tamper with the internal state via JSON export/reimport hack
      const entries = JSON.parse(logger.exportJSON());
      entries[0].action = 'TAMPERED';

      // Create a new logger and inject tampered entries
      const tampered = new AuditLogger();
      // Access private field for testing
      (tampered as any).entries = entries;

      expect(tampered.verify()).toBe(false);
    });
  });

  describe('getEntries()', () => {
    beforeEach(() => {
      logger.log({ agentId: 'agent-1', action: 'data.read', decision: 'allow' });
      logger.log({ agentId: 'agent-2', action: 'data.write', decision: 'deny' });
      logger.log({ agentId: 'agent-1', action: 'data.delete', decision: 'review' });
    });

    it('returns all entries without filter', () => {
      expect(logger.getEntries()).toHaveLength(3);
    });

    it('filters by agentId', () => {
      const entries = logger.getEntries({ agentId: 'agent-1' });
      expect(entries).toHaveLength(2);
      entries.forEach((e) => expect(e.agentId).toBe('agent-1'));
    });

    it('filters by action', () => {
      const entries = logger.getEntries({ action: 'data.write' });
      expect(entries).toHaveLength(1);
      expect(entries[0].decision).toBe('deny');
    });

    it('filters by date', () => {
      const past = new Date(Date.now() - 60_000);
      const entries = logger.getEntries({ since: past });
      expect(entries).toHaveLength(3);

      const future = new Date(Date.now() + 60_000);
      expect(logger.getEntries({ since: future })).toHaveLength(0);
    });
  });

  describe('exportJSON()', () => {
    it('exports valid JSON', () => {
      logger.log({ agentId: 'a', action: 'x', decision: 'allow' });
      const json = logger.exportJSON();
      const parsed = JSON.parse(json);
      expect(Array.isArray(parsed)).toBe(true);
      expect(parsed).toHaveLength(1);
    });
  });

  describe('maxEntries', () => {
    it('evicts old entries when limit is exceeded', () => {
      const small = new AuditLogger({ maxEntries: 3 });
      for (let i = 0; i < 5; i++) {
        small.log({ agentId: `a${i}`, action: 'x', decision: 'allow' });
      }
      expect(small.length).toBe(3);
    });
  });
});
